x/rewards/keeper/providers_test.go

```diff
@@ -0,0 +1,382 @@
+	err := ts.StakeProvider(providerAcc1.Addr.String(), ts.spec, testBalance)
+	require.Nil(t, err)
+
+	providerAcc2, _ := ts.AddAccount(common.PROVIDER, 2, 2*testBalance)
+	err = ts.StakeProvider(providerAcc2.Addr.String(), ts.spec, 2*testBalance)
+	require.Nil(t, err)
+
+	providerAcc3, _ := ts.AddAccount(common.PROVIDER, 3, 3*testBalance)
+	err = ts.StakeProvider(providerAcc3.Addr.String(), ts.spec, 3*testBalance)
+	require.Nil(t, err)
+
+	ts.AdvanceEpoch()
+
+	consumerAcc, _ := ts.AddAccount(common.CONSUMER, 1, ts.plan.Price.Amount.Int64())
+	_, err = ts.TxSubscriptionBuy(consumerAcc.Addr.String(), consumerAcc.Addr.String(), ts.plan.Index, 1, false)
+	require.Nil(t, err)
+
+	msg := ts.SendRelay(providerAcc1.Addr.String(), consumerAcc, []string{ts.spec.Index}, ts.plan.Price.Amount.Uint64()/2)
+	_, err = ts.TxPairingRelayPayment(msg.Creator, msg.Relays...)
+	require.Nil(t, err)
+
+	msg = ts.SendRelay(providerAcc2.Addr.String(), consumerAcc, []string{ts.spec.Index}, ts.plan.Price.Amount.Uint64())
+	_, err = ts.TxPairingRelayPayment(msg.Creator, msg.Relays...)
+	require.Nil(t, err)
+
+	msg = ts.SendRelay(providerAcc3.Addr.String(), consumerAcc, []string{ts.spec.Index}, ts.plan.Price.Amount.Uint64()*2)
+	_, err = ts.TxPairingRelayPayment(msg.Creator, msg.Relays...)
+	require.Nil(t, err)
+
+	// first months there are no bonus rewards, just payment ffrom the subscription
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAcc1.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	ts.TxDualstakingClaimRewards(providerAcc1.Addr.String(), "")
+
+	res, err = ts.QueryDualstakingDelegatorRewards(providerAcc2.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	ts.TxDualstakingClaimRewards(providerAcc2.Addr.String(), "")
+
+	res, err = ts.QueryDualstakingDelegatorRewards(providerAcc3.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	ts.TxDualstakingClaimRewards(providerAcc3.Addr.String(), "")
+
+	// now the provider should get all of the provider allocation
+	ts.AdvanceMonths(1)
+	distBalance := ts.Keepers.Rewards.TotalPoolTokens(ts.Ctx, types.ProviderDistributionPool)
+	ts.AdvanceEpoch()
+
+	res1, err := ts.QueryDualstakingDelegatorRewards(providerAcc1.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res1.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).SubRaw(3))
+
+	res2, err := ts.QueryDualstakingDelegatorRewards(providerAcc2.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res2.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).MulRaw(2))
+
+	res3, err := ts.QueryDualstakingDelegatorRewards(providerAcc3.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res3.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).MulRaw(4).AddRaw(6))
```

Comment:
add a comment the add 6 and sub 3 are because of remainder issues

Reply:
done